lib/rust/flexer/src/generate.rs

```diff
@@ -0,0 +1,526 @@
+                            reader.append_result(char);
+                            self.logger.info(||format!("Result is {:?}",reader.result()));
+                        },
+                        Err(flexer::prelude::reader::Error::EndOfGroup) => {
+                            let current_state = self.current_state();
+                            let group_name    = self.groups().group(current_state).name.as_str();
+                            let err           = format!("Missing rules for state {}", group_name);
+                            self.logger.error(err.as_str());
+                            panic!(err)
+                        }
+                        Err(_) => {
+                            self.logger.error("Unexpected error.");
+                            panic!("Unexpected error!")
+                        }
+                    }
+                    reader.advance_char(&mut self.bookmarks);
+                }
+            }
+
+            self.status
+        }
+    };
+    tree
+}
+
+/// Generate the `step` function for the lexer.
+///
+/// This function is responsible for dispatching based on the current state, consuming a character,
+/// and returning the state to transition to.
+pub fn step(groups:&group::Registry) -> ImplItem {
+    let arms = groups.all().iter().map(|g| step_match_arm(g.id.into())).collect_vec();
+    parse_quote! {
+        fn step<R:LazyReader>(&mut self, next_state:SubStateId, reader:&mut R) -> StageStatus {
+            let current_state:usize = self.current_state().into();
+            match current_state {
+                #(#arms)*
+                _ => unreachable_panic!("Unreachable state reached in lexer."),
+            }
+        }
+    }
+}
+
+/// Generate a match arm for the step function.
+///
+/// There is one match arm per lexer state.
+pub fn step_match_arm(number:usize) -> Arm {
+    let literal           = Literal::usize_unsuffixed(number);
+    let function_name_str = format!("dispatch_in_state_{}",number);
+    let func_name:Ident   = parse_str(function_name_str.as_str()).unwrap();
+    let arm:Arm = parse_quote! {
+        #literal => self.#func_name(next_state,reader),
+    };
+    arm
+}
+
+
+// === Generation for a Specific Lexer State ===
+
+/// Generate the functions that implement the lexer automaton for a given lexer state.
+pub fn automaton_for_group
+( group    : &Group
+, registry : &group::Registry
+) -> Result<Vec<ImplItem>,GenError> {
+    let nfa       = registry.to_nfa_from(group.id);
+    let mut rules = Vec::with_capacity(nfa.states.len());
+    for state in nfa.states.iter() {
+        if state.name.is_some() {
+            rules.push(rule_for_state(state)?);
+        }
+    }
+    let mut dfa             = DFA::from(&nfa);
+    let dispatch_for_dfa    = dispatch_in_state(&dfa,group.id.into())?;
+    let mut dfa_transitions = transitions_for_dfa(&mut dfa,group.id.into())?;
+    dfa_transitions.insert(0,dispatch_for_dfa);
```

Comment:
what is this 0? Please create a separate variable describing it. Its hard to understand it from this code context.